src/bsoncxx/test/bson_serialization.cpp

```diff
@@ -0,0 +1,322 @@
+// Copyright 2020 MongoDB Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <string>
+
+#include <bsoncxx/builder/basic/array.hpp>
+#include <bsoncxx/builder/basic/document.hpp>
+#include <bsoncxx/builder/stream/document.hpp>
+#include <bsoncxx/document/value.hpp>
+#include <bsoncxx/string/to_string.hpp>
+#include <bsoncxx/test_util/catch.hh>
+#include <bsoncxx/types/bson_value/view.hpp>
+
+using namespace bsoncxx;
+
+using builder::basic::kvp;
+using builder::basic::make_array;
+using builder::basic::make_document;
+
+namespace {
+
+namespace test {
+
+struct Person {
+    std::string first_name;
+    std::string last_name;
+    int age;
+};
+
+// to_bson serializer function that can take a document::value&
+void to_bson(const Person& person, bsoncxx::document::value& bson_object) {
+    bson_object = make_document(kvp("first_name", person.first_name),
+                                kvp("last_name", person.last_name),
+                                kvp("age", person.age));
+}
+
+void from_bson(Person& person, const bsoncxx::document::view& bson_object) {
+    bson_object["first_name"].get_value().to_field(person.first_name);
+    bson_object["last_name"].get_value().to_field(person.last_name);
+    bson_object["age"].get_value().to_field(person.age);
+}
+
+class Car {
+   public:
+    Car() = default;
+    Car(std::string manufacturer, std::string model)
+        : _manufacturer(std::move(manufacturer)), _model(std::move(model)){};
+
+    std::string get_manufacturer() const {
+        return _manufacturer;
+    }
+
+    std::string get_model() const {
+        return _model;
+    }
+
+    friend void to_bson(const Car& car, bsoncxx::document::value& bson_object) {
+        auto builder = bsoncxx::builder::stream::document{};
+        bson_object = builder << "manufacturer" << car._manufacturer << "model" << car._model
+                              << bsoncxx::builder::stream::finalize;
+    }
+
+    friend void from_bson(Car& car, const bsoncxx::document::view& bson_object) {
+        bson_object["manufacturer"].get_value().to_field(car._manufacturer);
```

Comment:
Per our discussion, I'm not sure what we'd gain by allowing the user to do this:
```
bson_object["manufacturer"].get_value().to_field(car._manufacturer);
```
instead of 
```
car._manufacturer = bson_object["manufacturer"].get_string().value.to_string();
```
They'd still have to consider the type of the BSON value and field to be set since `to_field` will attempt to assign to the object's field based on it's argument type. That last part of the process is now obfuscated by the method call. Consider this case:
```
int i = 32;
bson_object["manufacturer"].get_value().to_field(i);
``` 
This will compile and set `i` to `"Tesla"` which results in a garbage value. So there isn't additional error checking. Had the user assigned a value like so:
```
int i = 32;
i = bson_object["manufacturer"].get_string().value.to_string();
```
They would've gotten a compile-time error. The `to_field` method would also have to handle rvalues, e.g., neither of these code snippets compile:
```
bson_object["manufacturer"].get_value().to_field("hello");
...
bson_object["manufacturer"].get_value().to_field(32);
```
I think we can remove the `to_field` function for now.

Reply:
I agree that the way `to_field` is implemented right now leaves very little to gain when the user can simply just assign the values from the BSON object to their own object themselves and it'd be best to remove it. I would like to address your examples though.

For the example in which we use `to_field(i)` where `i` is an `int`, we could change the `to_field` overloads to check for type first before assigning:
```
_to_field(int32_t& object_field) const {
    if (_type == bsoncxx::type::k_int32) {
        object_field = _b_int32.value;
    }
}
```
This would prevent the user's field from getting the incorrect value (i would remain 32 in this case). However, the user would definitely still need to pay attention to the types that they are using and may offer less flexibility, especially when it comes to implicit conversions. (For example, having an `int64_t i` while trying to obtain the value of a `b_int32` element would be impossible using `to_field` unless we make the function check that type as well.)

I am curious about your thoughts about the rvalues examples. Don't we want the code snippets to fail to compile when passing rvalues to the function since the assumed precondition is that we are using `to_field` to assign to an lvalue (member field)? The user shouldn't be doing something like `"hello" = bson_object...` to begin with.

Reply:
Ah yes! Ignore the rvalue comments. It should fail in that case. Regarding the type checking, it's _possible_ but it would require a lot of work on our part to function correctly. It also isn't very idiomatic of C++ to provide functions solely for assignment. At least I haven't seen anything like it in the standard library. Usually, if you want to assign to a variable you have the implicit and explicit constructors, copy constructors, and move constructors, converting constructors, etc (see [this](https://en.cppreference.com/w/cpp/language/converting_constructor) for more examples).